File: examples/guests/python/decoder/decoder.py
from generated.guest.contracts import (
    DECODERPipelineDecoderPlugin,
    set_decoder_factory,
    polyplug_init,
)


class DecoderImpl(DECODERPipelineDecoderPlugin):
    """The factory receives the HostApi pointer at polyplug_init time."""

    def __init__(self, host_ptr: int) -> None:
        # Host handle for this runtime, captured at construction.
        self._host_ptr: int = host_ptr

    def decode(self, input: str) -> str:
        return f"DECODED:{input.replace(',', '|')}"


# Register the factory; the generated polyplug_init constructs the
# implementation with its owning runtime's host pointer.
set_decoder_factory(DecoderImpl)

File: examples/guests/csharp/decoder/Decoder.cs
using System.Runtime.CompilerServices;
using Polyplug.Guest;
using Polyplug.Abi;

namespace Decoder;

public sealed class DecoderPlugin : IPipelineDecoderGuestContract
{
    // Host handle for this runtime, captured at instance creation.
    private readonly IntPtr _host;

    public DecoderPlugin(IntPtr host)
    {
        _host = host;
    }

    public StringView Decode(StringView input)
    {
        string s = StringViewHelper.ToString(input).Replace(',', '|');
        return PolyplugHost.AllocString(_host, $"DECODED:{s}");
    }
}

public static class Registration
{
    [ModuleInitializer]
    public static void Register()
    {
        // The factory receives the HostApi pointer per created instance.
        DecoderInterfaces.SetDecoderFactory(host => new DecoderPlugin(host));
    }
}
